tokio/src/io/util/vec_with_initialized.rs

```diff
@@ -47,19 +48,28 @@ where
         // to its length are initialized.
         Self {
             num_initialized: vec.as_mut().len(),
+            starting_capacity: vec.as_ref().capacity(),
             vec,
         }
     }
 
-    pub(crate) fn reserve(&mut self, num_bytes: usize) {
+    // Returns a boolean telling the caller to try reading into a small local buffer first if true.
+    // Doing so would avoid overallocating when vec is filled to capacity and we reached EOF.
+    pub(crate) fn reserve(&mut self, num_bytes: usize) -> bool {
```

Comment:
I wouldn't put this logic in the `reserve` method. I think that would be confusing.